src/primitives/common/scanline.rs

```diff
@@ -0,0 +1,175 @@
+use core::ops::Range;
+
+use crate::{
+    geometry::{Point, Size},
+    prelude::Primitive,
+    primitives::{Line, Rectangle},
+};
+
+/// Scanline.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct Scanline {
+    pub y: i32,
+    pub x: Range<i32>,
+}
+
+impl Scanline {
+    /// Creates a new empty scanline.
+    pub const fn new(y: i32) -> Self {
+        Self { y, x: 0..0 }
+    }
+
+    /// Returns `true` if the x range of the scanline is empty.
+    pub fn is_empty(&self) -> bool {
+        // MSRV: use `Range::is_empty` on version >= 1.47.0
+        !(self.x.start < self.x.end)
+    }
+
+    /// Extends the scanline to include the given x coordinate.
+    fn extend(&mut self, x: i32) {
+        if self.is_empty() {
+            self.x = x..x + 1;
+        } else {
+            if x < self.x.start {
+                self.x.start = x;
+            } else if x >= self.x.end {
+                self.x.end = x + 1;
+            }
+        }
+    }
+
+    /// Intersect a horizontal scan line with the Bresenham representation of this line segment.
+    ///
+    /// Intersection lines produced by this function are sorted so that the start always lies to the
+    /// left of the end.
+    pub fn bresenham_intersection(&mut self, line: &Line) {
+        // Check if the scanline is in the y range of the line.
+        let y_range = if line.start.y <= line.end.y {
+            line.start.y..=line.end.y
+        } else {
+            line.end.y..=line.start.y
+        };
+
+        if !y_range.contains(&self.y) {
+            return;
+        }
+
+        let y = self.y;
+        let mut points = line.points();
+
+        // Find the first point with the same y coordinate.
+        while let Some(first) = points.find(|p| p.y == y) {
+            self.extend(first.x);
+        }
+
+        // Check points until the y coordinate changes.
+        if let Some(last) = points.take_while(|p| p.y == y).last() {
+            self.extend(last.x);
+        }
```

Comment:
This could be shrunk down to:

```suggestion
        line.points()
            .skip_while(|p| p.y != y)
            .take_while(|p| p.y == y)
            .for_each(|p| {
                self.extend(p.x);
            });
```

I was hoping Rust would be able to optimise an iterator a bit better but there's no change on top of the benchmarks you already posted. At least it's more concise :)

Reply:
You code has a slightly different behavior. My version did only call `extend` for the first and last pixel with the scanline y position and your version does call `extend` for every pixel with the same y coordinate. This would only make a performance difference for lines that are (almost) horizontal, so maybe it's better to just call `extend` for every pixel to be sure the range is correctly calculated in all cases. And your code looks much nicer.

Reply:
Ah, yes you're right. I think it's probably better to use your version in that case. There might be a performance cliff for nearly horizontal lines which would be nice to avoid.

Reply:
I just tested your version and it was a bit faster than the old one, nice.

Reply:
Did you test with nearly-horizontal long lines? I'd be interested to see if that's any quicker or slower.

Reply:
No, I haven't, but that would indeed be interesting. I don't have time at the moment, so if you want to give it a shot go ahead, otherwise I'll test it later.

Reply:
I ran a quick scrappy benchmark locally and the iterator approach doesn't seem to change performance. I see a ~+2% speedup in `thick polyline` but that's so small it's not really worth mentioning.